### PR TITLE
feat: add Roberts Cross edge detection operator

### LINKED ISSUE #195: [Feature] Add Roberts Cross edge detection operator
**Difficulty:** Easy
**Area:** full stack

## Description
The `sobel_derivatives` package ships Sobel, Scharr and Prewitt but no Roberts Cross, so the classical edge-detector set is incomplete on a platform whose point is letting students compare gradient operators side by side. OpenCV has no Roberts builtin, so it is implemented the same way `PrewittOperator` already is: fixed kernels pushed through `cv2.filter2D` into `CV_64F`, combined with `np.hypot`, then clipped back to uint8. Roberts Cross uses the two 2x2 diagonal kernels `[[1, 0], [0, -1]]` and `[[0, 1], [-1, 0]]`. `prewitt_operator.py` is a direct template for the whole operator, including its uint8 dtype guard and its 1/3/4 channel handling, so this is a mechanical addition across the usual six registration points.

## Acceptance Criteria
- [ ] `imagelab-backend/app/operators/sobel_derivatives/roberts_operator.py` defines a `RobertsOperator` mirroring `PrewittOperator` (uint8 guard, 1/3/4 channel handling, `np.hypot` magnitude, clip to 255)
- [ ] The operator is registered in `OPERATOR_REGISTRY` under a `sobelderivatives_` key
- [ ] A Blockly definition is added to `sobel-derivatives.blocks.ts` using `style: "sobel_derivatives_style"` with `previousStatement` and `nextStatement` set
- [ ] The block is listed under the Sobel Derivatives category in `categories.ts`
- [ ] An entry is added to `operatorDocs.ts` so the block has in-app documentation
- [ ] Unit tests under `imagelab-backend/tests/operators/sobel_derivatives/` cover a synthetic edge image, a grayscale input, and rejection of a non-uint8 array
- [ ] `uv run pytest -v` and the frontend `npm run lint` both pass

## Resources
- `imagelab-backend/app/operators/sobel_derivatives/prewitt_operator.py:6` - the module-level immutable kernels, and the `compute` body from line 21 that this operator should mirror
- `imagelab-backend/app/operators/registry.py:125` - where `sobelderivatives_prewittoperator` is registered
- `imagelab-frontend/src/blocks/definitions/sobel-derivatives.blocks.ts:42` - the Prewitt block definition to copy
- `imagelab-frontend/src/blocks/categories.ts:139` - the Sobel Derivatives category listing
- `imagelab-frontend/src/data/operatorDocs.ts:426` - the Prewitt documentation entry
- `imagelab-backend/tests/operators/sobel_derivatives/test_scharr_derivative.py` - existing test layout for this package
- [OpenCV filter2D tutorial](https://docs.opencv.org/4.x/d4/dbd/tutorial_filter_2d.html)

<details>
<summary>Original report by @xxivani (2026-03-12)</summary>

**Is your feature request related to a problem?**

ImageLab currently has Sobel and Scharr edge detection operators but is missing the Roberts Cross operator. For a teaching platform, having the full set of classical edge detectors is valuable for demonstrating how different gradient operators compare.

**Describe the solution you'd like**

A new Roberts Cross operator in the Sobel Derivatives category that applies the two 2x2 diagonal kernels to compute gradient magnitude, following the same full-stack pattern as the existing Sobel and Scharr operators.

**Describe alternatives you've considered**

Roberts Cross is not natively implemented in OpenCV so it is implemented manually using cv2.filter2D with the two 2x2 kernels, which is the standard approach.

**Additional context**
Any other context or screenshots.

</details>

### PR DESCRIPTION
**Pull request by xxivani**
## Description
Adds a Roberts Cross edge detection operator to the Sobel Derivatives category. The backend applies two 2x2 diagonal gradient kernels (`Gx = [[1,0],[0,-1]]`, `Gy = [[0,1],[-1,0]]`) via `cv2.filter2D` and returns the gradient magnitude as a uint8 image. Colour images are automatically converted to grayscale. 

Fixes #195 

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update

## How Has This Been Tested?
Ran the pipeline with a Roberts Cross block on both grayscale and colour images and confirmed correct edge detection output.

- [x] Existing tests pass
- [ ] New tests added
- [x] Manual testing

## Checklist
- [x] My code follows the project's style guidelines
- [x] I have performed a self-review
- [x] I have added/updated documentation as needed
- [x] My changes generate no new warnings
- [x] Tests pass locally

### CODE REVIEW

File: imagelab-backend/app/operators/sobel_derivatives/roberts_cross.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    def compute(self, image: np.ndarray) -> np.ndarray:
+        if image.dtype != np.uint8:
+            raise ValueError(f"RobertsCross expects a uint8 image, got dtype={image.dtype}.")
+        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY) if len(image.shape) == 3 else image
+
+        grad_x = cv2.filter2D(gray, cv2.CV_64F, _KERNEL_X)
+        grad_y = cv2.filter2D(gray, cv2.CV_64F, _KERNEL_Y)
+
+        magnitude = np.sqrt(grad_x**2 + grad_y**2)
+        return np.clip(magnitude, 0, 255).astype(np.uint8)
```
Comment (ivantha) [inline imagelab-backend/app/operators/sobel_derivatives/roberts_cross.py]: **[HIGH]** No unit tests for the new operator

The PR explicitly acknowledges that no new tests were added. The two runtime-crash bugs identified above would have been caught immediately by even minimal unit tests. The existing codebase appears to have a test suite (evidenced by the checklist item 'Existing tests pass'), but `RobertsCross` is not covered at all.

Missing test scenarios:
- Grayscale `(H, W)` input
- BGR `(H, W, 3)` input (colour-to-gray path)
- `(H, W, 1)` single-channel-axis input (currently crashes)
- `(H, W, 4)` BGRA input (currently crashes or wrong)
- Non-`uint8` input (e.g. `float32`) → expected `ValueError`
- Known output values on a synthetic 4×4 test image

**Suggestion:**
```
Add a test file `tests/operators/sobel_derivatives/test_roberts_cross.py`:
```python
import numpy as np
import pytest
from app.operators.sobel_derivatives.roberts_cross import RobertsCross

op = RobertsCross()

def make(shape, dtype=np.uint8):
    return np.zeros(shape, dtype=dtype)

def test_grayscale_2d():
    img = make((10, 10))
    out = op.compute(img)
    assert out.shape == (10, 10)
    assert out.dtype == np.uint8

def test_bgr_3channel():
    img = make((10, 10, 3))
    out = op.compute(img)
    assert out.shape == (10, 10)

def test_bgra_4channel():
    img = make((10, 10, 4))
    out = op.compute(img)
    assert out.shape == (10, 10)

def test_single_channel_axis():
    img = make((10, 10, 1))
    out = op.compute(img)
    assert out.shape == (10, 10)

def test_non_uint8_raises():
    img = make((10, 10), dtype=np.float32)
    with pytest.raises(ValueError):
        op.compute(img)

def test_known_output():
    # A 4x4 image with a sharp diagonal step
    img = np.array([[0,0,255,255],[0,0,255,255],
                    [0,0,255,255],[0,0,255,255]], dtype=np.uint8)
    out = op.compute(img)
    assert out.max() > 0  # edges must be detected
```
```

File: imagelab-backend/app/operators/sobel_derivatives/roberts_cross.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+_KERNEL_Y = np.array([[0, 1], [-1, 0]], dtype=np.float64)
+
+
+class RobertsCross(BaseOperator):
+    """Applies the Roberts Cross operator to detect edges using diagonal gradients."""
+
+    def compute(self, image: np.ndarray) -> np.ndarray:
+        if image.dtype != np.uint8:
+            raise ValueError(f"RobertsCross expects a uint8 image, got dtype={image.dtype}.")
+        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY) if len(image.shape) == 3 else image
```
Comment (ivantha) [inline imagelab-backend/app/operators/sobel_derivatives/roberts_cross.py]: **[MEDIUM]** Hardcoded BGR color-space assumption may be wrong for the pipeline

The conversion uses `cv2.COLOR_BGR2GRAY`, which implicitly assumes the upstream pipeline stores colour images in BGR order. If the application stores images in RGB order (common when images come from PIL/Pillow, browsers, or are decoded with `imageio`), the R and B channels are swapped during the luminance calculation.

OpenCV's BGR-to-gray formula is `Y = 0.114·B + 0.587·G + 0.299·R`, so swapping R and B slightly changes the grayscale result. While the visual difference is subtle and edges are largely unaffected, the behaviour is inconsistent with user expectations and may diverge from how other operators in the pipeline handle the same conversion.

**Suggestion:**
```
Check how other operators in the `sobel_derivatives` category perform the BGR-vs-RGB conversion and use the same convention. If the codebase uses RGB internally, change to `cv2.COLOR_RGB2GRAY`.
```

File: imagelab-backend/app/operators/sobel_derivatives/roberts_cross.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+_KERNEL_X = np.array([[1, 0], [0, -1]], dtype=np.float64)
+_KERNEL_Y = np.array([[0, 1], [-1, 0]], dtype=np.float64)
+
+
+class RobertsCross(BaseOperator):
+    """Applies the Roberts Cross operator to detect edges using diagonal gradients."""
+
+    def compute(self, image: np.ndarray) -> np.ndarray:
+        if image.dtype != np.uint8:
+            raise ValueError(f"RobertsCross expects a uint8 image, got dtype={image.dtype}.")
```
Comment (ivantha) [inline imagelab-backend/app/operators/sobel_derivatives/roberts_cross.py]: **[MEDIUM]** Over-strict dtype guard rejects valid float inputs without a normalisation path

Many image processing pipelines operate on `float32` or `float64` images (e.g. after HDR loading, after a previous operator outputs floats). The hard `raise ValueError` for any non-`uint8` dtype means `RobertsCross` cannot participate in such pipelines at all, even though the underlying `cv2.filter2D` call would handle float input correctly.

This is more restrictive than is necessary, and it is inconsistent if peer operators accept float images.

**Suggestion:**
```
Either (a) accept float images by normalising to `[0, 255]` before processing, or (b) add a conversion step:
```python
if image.dtype != np.uint8:
    # Normalise to uint8 so downstream logic is uniform
    norm = cv2.normalize(image, None, 0, 255, cv2.NORM_MINMAX)
    image = norm.astype(np.uint8)
```
Alternatively, keep the strict check but document it clearly in the docstring so callers know to convert beforehand.
```

File: imagelab-backend/app/operators/sobel_derivatives/roberts_cross.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY) if len(image.shape) == 3 else image
+
+        grad_x = cv2.filter2D(gray, cv2.CV_64F, _KERNEL_X)
+        grad_y = cv2.filter2D(gray, cv2.CV_64F, _KERNEL_Y)
```
Comment (ivantha) [inline imagelab-backend/app/operators/sobel_derivatives/roberts_cross.py]: **[LOW]** Implicit anchor for 2×2 kernel may shift gradient by one pixel

The `cv2.filter2D` calls use the default `anchor=(-1, -1)`, which OpenCV resolves to `(kernel_height//2, kernel_width//2)`. For a **2×2 kernel** this evaluates to anchor `(1, 1)` — the bottom-right corner — rather than `(0, 0)` (top-left), which is the canonical anchor for the Roberts Cross operator.

With anchor `(1, 1)`, `_KERNEL_X` computes `I[r-1, c-1] − I[r, c]` instead of the traditional `I[r, c] − I[r+1, c+1]`. Both expressions measure the same diagonal gradient but the result is assigned to a position shifted one pixel up-left. For gradient *magnitude* the effect is cosmetically negligible, but it could matter if gradient direction or precise spatial alignment is ever needed downstream.

**Suggestion:**
```
Explicitly pass `anchor=(0, 0)` to make the intent unambiguous and match the classical Roberts Cross definition:
```python
grad_x = cv2.filter2D(gray, cv2.CV_64F, _KERNEL_X, anchor=(0, 0))
grad_y = cv2.filter2D(gray, cv2.CV_64F, _KERNEL_Y, anchor=(0, 0))
```
```

File: imagelab-backend/app/operators/sobel_derivatives/roberts_cross.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import cv2
+import numpy as np
+
+from app.operators.base import BaseOperator
+
+# Roberts Cross kernels — 2x2 diagonal gradient operators
+_KERNEL_X = np.array([[1, 0], [0, -1]], dtype=np.float64)
+_KERNEL_Y = np.array([[0, 1], [-1, 0]], dtype=np.float64)
```
Comment (ivantha) [inline imagelab-backend/app/operators/sobel_derivatives/roberts_cross.py]: **[INFO]** Module-level kernel constants are a clean and efficient pattern

Defining `_KERNEL_X` and `_KERNEL_Y` at module level (private, typed `float64`) avoids re-allocating NumPy arrays on every `compute()` call. The leading underscore correctly signals that these are implementation details. This is a positive pattern worth preserving.

**Suggestion:**
```
No change needed. This is good practice.
```

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

The PR cleanly integrates the Roberts Cross operator into the existing pattern (registry entry, block definition, frontend category, operator docs) and the core algorithm — two 2×2 diagonal `filter2D` passes followed by Euclidean magnitude — is mathematically correct. However, two hard-crash bugs exist in the channel-detection logic: a `(H, W, 1)` grayscale image and any 4-channel BGRA/RGBA image will both trigger `cv2.error` at runtime because `len(image.shape) == 3` is True for all three-dimensional arrays regardless of channel count, and `cv2.COLOR_BGR2GRAY` only accepts 3-channel input. The absence of any automated tests means these bugs were not caught before submission. Fixing the channel-count guard and adding a small unit test suite are the two must-haves before merging; the remaining findings (implicit filter anchor, strict dtype check, missing docstrings) are lower priority improvements.

**Suggested Version Bump:** `minor`

### ⚠️ Breaking Changes

- **[INFO]** No breaking changes — purely additive feature: **Previous behavior**: The `sobelderivatives` category contained only `SobelDerivative` and `ScharrDerivative`.

**New behavior**: A third operator, `RobertsCross`, is registered under the key `sobelderivatives_robertscross` and exposed in the frontend block palette.

**Who is affected**: No existing consumers are affected. The new registry key is unique and does not shadow or replace any existing key.

**Migration path**: None required — existing pipelines are unaffected.
  > **Suggestion:** No action needed.


### PR Description Issues

- **[LOW]** Screenshots section is empty: The PR description includes a '## Screenshots (if applicable)' heading with no content beneath it. For an operator that produces visual output (edge-detected images), before/after screenshots would be valuable for reviewers to visually verify correctness — especially since manual testing is the primary validation method cited.
  > **Suggestion:** Add at least one before/after screenshot pair, e.g.:
```
## Screenshots
| Input | Roberts Cross output |
|-------|---------------------|
| ![input](url) | ![output](url) |
```
Alternatively, remove the heading entirely if screenshots are intentionally omitted.
- **[LOW]** Testing section acknowledges no new tests — should clarify rationale or add tests: The PR description checks 'Existing tests pass' and 'Manual testing' but explicitly leaves 'New tests added' unchecked. Given that the new operator introduces a new code path with non-trivial edge cases (RGBA input, single-channel-axis input, dtype validation), the absence of automated tests is a gap that should either be addressed or explicitly justified (e.g. referencing a follow-up ticket).
  > **Suggestion:** Either add unit tests (see the testing finding above for a starter suite), or add a note in the PR description explaining why tests are deferred and referencing a follow-up issue:
> _Tests will be added in #XXX as part of the broader operator test coverage initiative._

**Comment by ivantha** [comment]
Reviewed. Adds a Roberts Cross edge detection operator with a clean full-stack implementation — backend operator using 2×2 diagonal kernels with proper grayscale conversion and uint8 normalization, registry registration, frontend block definition, category listing, and `operatorDocs` entry with formula. Test coverage includes shape assertions, float input normalization, uniform-image zero-output, step-edge detection, and immutability. Merging via rebase.

**Comment by ivantha** [comment]
Thanks for the PR! It currently has merge conflicts with main (or is behind). Could you rebase onto the latest main and push the updated branch? Happy to merge once it's auto-rebasable.

**Comment by xxivani** [comment]
Hi @ivantha, I've pushed the updated branch!

**Comment by ivantha** [comment]
Thanks for the contribution! This PR became not-mergeable after recent merges shifted `main`. Closing as part of a backlog cleanup — if you'd still like this merged, please rebase on the latest `main` and open a fresh PR. Your work is appreciated.
